jabber-mam.el

```diff
@@ -0,0 +1,206 @@
+               (when (and mam-stamp mam-msg-dir mam-msg-from mam-msg-to
+                          mam-msg-body-txt)
+                 (push (vector mam-stamp mam-msg-dir mam-msg-from
+                               mam-msg-to mam-msg-body-txt)
+                       jabber-mam-results))))))
+        ((jabber-xml-get-children xml-data 'fin)
+         ;; End of set, determine if a subsequent query is required (if the
+         ;; result is not complete).
+         ;; Extract "complete" attribute from <fin> tag and <last> id
+         (let* ((fin (jabber-xml-get-children xml-data 'fin))
+                (complete (jabber-xml-get-attribute (car fin) 'complete))
+                (set (jabber-xml-get-children (car fin) 'set))
+                (last
+                 (when set (jabber-xml-get-children (car set) 'last)))
+                (last-id (when last (cadr (cdr (car last))))))
+           (if (and (or (not complete) (not (string= complete "true"))) last-id)
+               ;; Result set is not complete, next request should start with
+               ;; `last-id'
+               (setq jabber-mam-last-id last-id)
+             ;; Result set is complete
+             (setq jabber-mam-done t))
+           ;; Release lock
+           (setq jabber-mam-lock t)
+           nil))
+        (t nil)))
+
+(defun jabber-mam-query (jc jid-me jid-with start-time end-time number
+                            direction)
+  "Build and send MAM query to server.
+JC is jabber connection.  Messages between users with JIDs JID-ME JID and
+JID-WITH JID with timestamp between START-TIME and END-TIME are retrieved.  The
+set of results is limited to NUMBER messages.  DIRECTION is either \"in\" or
+\"out\", or t for no limit on direction (this parameter is currently ignored)."
+
+  ;; Initialize output and lock
+  (setq jabber-mam-results (list))
+  (setq jabber-mam-done nil)
+  (setq jabber-mam-last-id nil)
+  (let ((number-left (if (integerp number) number nil)))
+    (while (not jabber-mam-done)
+      (let ((mam-query (jabber-mam-make-query
+                        jid-me jid-with
+                        start-time
+                        end-time
+                        number-left
+                        jabber-mam-last-id)))
+        ;;(message "MAM request: [%s]" (jabber-sexp2xml mam-query))
+        (setq jabber-mam-lock nil)
+        (jabber-send-iq jc nil "set" mam-query
+                        #'jabber-report-success "MAM request"
+                        #'jabber-report-success "MAM request")
+        ;; Wait for results
+        (while (not jabber-mam-lock)
+          (sit-for 1))
+        (when (integerp number)
+          (setq number-left (- number (length jabber-mam-results)))
+          (setq jabber-mam-done (or jabber-mam-done
+                                    (<= number-left 0)))))))
+  ;;(message "MAM got %d messages" (length jabber-mam-results))
+  (nreverse jabber-mam-results))
+
+(provide 'jabber-mam)
+;;; jabber-mam.el ends here
```

Comment:
Functions in these files are quite long. Your code would gain readability if some smaller functions were extracted from the large ones.